### PR TITLE
[stdlib] Have Substring.filter return a String

### PR DESCRIPTION
**Pull request by airspeedswift**
Per guidance in SE-163 that methods that return new strings shouldn't return substrings, `Substring` shouldn't just take the default for `RangeReplaceableCollection.filter` of returning `Self`.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

File: test/stdlib/subString.swift [hunk 1]
```diff
@@ -92,6 +92,14 @@ SubstringTests.test("Comparison") {
 		["apple", "pen", "pen", "pineapple"])
 }
 
+SubstringTests.test("Filter") {
+	var name = "😂Edward Woodward".dropFirst()
```
Comment (CodaFi) [inline test/stdlib/subString.swift]: Nit: Spaces here

**Comment by dabrahams** [comment]
I don't see the guidance you mention in SE-0163, but maybe I've overlooked something.  Could you clarify?

**Comment by timvermeulen** [comment]
Probably referring to this:

>The slicing operations on `String` will be amended to return `Substring`:
>
>```
>struct String {
>   subscript(bounds: Range<Index>) -> Substring { ... }
>}
>```
>
>Note that properties or methods that due to their nature create new `String` storage (such as `lowercased()`) will not change.

**Comment by dabrahams** [comment]
@timvermeulen this is not an existing method on String, but a new method (due to collection conformance) on a new type. Actually this creates an overload. i don't honk that phrase justifies this change.

**Comment by airspeedswift** [comment]
Tim’s correct.  There isn’t a good reason to return a Substring here. Similarly, when Max implemented the RRC.filter, he made sure the filter on ArraySlice remained returning an Array (which is why he raised https://bugs.swift.org/browse/SR-4933)

File: stdlib/public/core/Substring.swift.gyb [hunk 2]
```diff
@@ -428,6 +428,12 @@ extension Substring {
   public func uppercased() -> String {
     return String(self).uppercased()
   }
+  
+  public func filter(
+  _ isIncluded: (Element) throws -> Bool
```
Comment (moiseev) [inline stdlib/public/core/Substring.swift.gyb]: Not indented.

File: stdlib/public/core/Substring.swift.gyb [hunk 3]
```diff
@@ -428,6 +428,12 @@ extension Substring {
...
+  ) rethrows -> String {
+      return try String(self.lazy.filter(isIncluded))
```
Comment (moiseev) [inline stdlib/public/core/Substring.swift.gyb]: Indented with 4 spaces.

**Review by moiseev — COMMENTED** [review]
My quick repl sessions shows that both `String` and `Substring` call the `Sequence.filter` that returns `[Character]`. If that is correct than this change only solves half the problem (given there is a problem).

**Comment by dabrahams** [comment]
on Tue Jul 11 2017, Ben Cohen <notifications-AT-github.com> wrote:

> Tim’s correct.  There isn’t a good reason to return a Substring
> here. 

Do you want to know the reasons I can think of?

> Similarly, when Max implemented the RRC.filter, he made sure the
> filter on ArraySlice remained returning an Array

That was done for backward compatibility reasons and quite possible
should have been marked with availability to make it go away for Swift
4.

> (which is why he raised https://bugs.swift.org/browse/SR-4933)

Sorry, that link is dead for me.

-- 
-Dave

**Comment by airspeedswift** [comment]
@swift-ci please smoke test

**Pull request merged**